pkg/velerorunner/registry.go

```diff
@@ -0,0 +1,526 @@
+	labels["app"] = name
+	deploymentconfig := &appsv1API.DeploymentConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:    labels,
+			Name:      name,
+			Namespace: VeleroNamespace,
+		},
+	}
+	err := t.updateRegistryDeploymentConfig(deploymentconfig, name, srcName)
+	return deploymentconfig, err
+}
+
+// Update a Registry DeploymentConfig as desired for the specified cluster.
+func (t *Task) updateRegistryDeploymentConfig(deploymentconfig *appsv1API.DeploymentConfig, name, srcName string) error {
+	deploymentconfig.Spec = appsv1API.DeploymentConfigSpec{
+		Replicas: 1,
+		Selector: map[string]string{
+			"app":              name,
+			"deploymentconfig": name,
+		},
+		Strategy: appsv1API.DeploymentStrategy{Resources: kapi.ResourceRequirements{}},
+		Template: &kapi.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				CreationTimestamp: metav1.Time{},
+				Labels: map[string]string{
+					"app":              name,
+					"deploymentconfig": name,
+				},
+			},
+			Spec: kapi.PodSpec{
+				Containers: []kapi.Container{
+					kapi.Container{
+						Env: []kapi.EnvVar{
+							kapi.EnvVar{
+								Name:  "REGISTRY_STORAGE",
+								Value: "s3",
+							},
+							kapi.EnvVar{
+								Name: "REGISTRY_STORAGE_S3_ACCESSKEY",
+								ValueFrom: &kapi.EnvVarSource{
+									SecretKeyRef: &kapi.SecretKeySelector{
+										LocalObjectReference: kapi.LocalObjectReference{Name: name},
+										Key:                  "access_key",
+									},
+								},
+							},
+							kapi.EnvVar{
+								Name:  "REGISTRY_STORAGE_S3_BUCKET",
+								Value: t.PlanResources.MigStorage.Spec.BackupStorageConfig.AwsBucketName,
+							},
+							kapi.EnvVar{
+								Name:  "REGISTRY_STORAGE_S3_REGION",
+								Value: t.PlanResources.MigStorage.Spec.BackupStorageConfig.AwsRegion,
+							},
+							kapi.EnvVar{
+								Name:  "REGISTRY_STORAGE_S3_ROOTDIRECTORY",
+								Value: "/" + srcName,
+							},
+							kapi.EnvVar{
+								Name: "REGISTRY_STORAGE_S3_SECRETKEY",
+								ValueFrom: &kapi.EnvVarSource{
+									SecretKeyRef: &kapi.SecretKeySelector{
+										LocalObjectReference: kapi.LocalObjectReference{Name: name},
+										Key:                  "secret_key",
+									},
+								},
+							},
+						},
+						Image: "registry:2",
+						Name:  name,
+						Ports: []kapi.ContainerPort{
+							kapi.ContainerPort{
+								ContainerPort: 5000,
+								Protocol:      kapi.ProtocolTCP,
+							},
+						},
+						Resources: kapi.ResourceRequirements{},
+						VolumeMounts: []kapi.VolumeMount{
+							kapi.VolumeMount{
+								MountPath: "/var/lib/registry",
+								Name:      name + "-volume-1",
+							},
+						},
+					},
+				},
+				Volumes: []kapi.Volume{
+					kapi.Volume{
+						Name:         name + "-volume-1",
+						VolumeSource: kapi.VolumeSource{EmptyDir: &kapi.EmptyDirVolumeSource{}},
+					},
+				},
+			},
+		},
+		Test: false,
+		Triggers: appsv1API.DeploymentTriggerPolicies{
```

Comment:
What's going on with this DeploymentTrigger?

Reply:
I'm basically trying to recreate, as closely as possible, the deployment config definition returned by `oc new-app`. For the registry creation that worked in my dev environment, took the output produced by `oc new-app <...> --dry-run -o yaml` and reproduced it here. If there are things you think I should remove that would leave this fully-functional, I could do that, but it seemed safer as a starting point to replicate what the oc tool did.

Reply:
But to answer your question more specifically, if I'm understanding dc triggers properly, this should trigger a new pod any time the dc config changes or the associated imagestream gets a new image.

Reply:
Ok. I imagine we won't be publishing new stuff to the imagestream that would trigger a redeploy. Was thinking we don't want the registry to go down during a migration.

Reply:
We could pull it out, since I imagine this should be a pretty stable codebase anyway -- no reason we should be updating the underlying image for the registry, although we should be careful about the change. The registry will be up from the first stage migration until the final migrate. Will we eventually support controller upgrades with in-progress plans? Hopefully we could limit that -- basically if we can assume that we will never update the controller code between stage and migrate, then we can probably eliminate the image trigger. We probably want to keep the config trigger though.

Reply:
Yeah I think you both bring up good points. I would leave the config trigger and remove the build trigger since config changes are not likely to occur during the migration run whereas a build trigger I could see the problem arising. But honestly, even if you leave the build trigger I don't think it would cause any problems, just more to be safe. OCP should scale up a pod before taking the old one down.